chain-core/src/state/account.rs

```diff
@@ -237,30 +269,40 @@ impl StakedState {
     /// Checks if current account is jailed
     #[inline]
     pub fn is_jailed(&self) -> bool {
-        self.jailed_until.is_some()
+        self.punishment.is_some()
     }
 
     /// Returns `jailed_until` for current account, `None` if current account is not jailed
     #[inline]
     pub fn jailed_until(&self) -> Option<Timespec> {
-        self.jailed_until
+        self.punishment
+            .as_ref()
+            .map(|punishment| punishment.jailed_until)
     }
 
     /// Jails current account until given time
-    pub fn jail_until(&mut self, jail_until: Timespec) {
+    pub fn jail_until(&mut self, jailed_until: Timespec, kind: PunishmentKind) {
         self.nonce += 1;
-        self.jailed_until = Some(jail_until)
+        self.punishment = Some(Punishment {
+            kind,
+            jailed_until,
+            slash_amount: None,
+        });
     }
 
     /// Unjails current account
     pub fn unjail(&mut self) {
         self.nonce += 1;
-        self.jailed_until = None;
+        self.punishment = None;
```

Comment:
In order to do this, it should be enforced `slash_wait_period < jail_duration`, e.g. in `validate_config_get_genesis`